_posts/2020-11-30-flexible-and-safe-tuples-in-scala3.md

```diff
@@ -0,0 +1,385 @@
+and program against.
+
+## New operations
+
+A lot of operations are now available on tuples out of the box!
+
+Many of these were possible only using third-party libraries such
+as Shapeless in Scala 2, which was a complicated task for new
+Scala developers.
+
+These operations are now available in the standard library, they
+are safe and preserve the individual types of each element.
+
+The first one was already introduced: `.size` retrieves the number
+of elements in the tuple.
+
+Let's see:
+
+### Tuples can grow
+
+We can add an element to a tuple using the `*:` operator,
+which is very similar to the `::` combinator available on `List`.
+
+```scala
+val fourWeirdNumbers = (1, "2", 3.0, List(4))
+val evenWeirder = 1 *: "2" *: 3.0 *: List(4) *: Tuple()
+
+val thisIsTrue = fourWeirdNumbers == evenWeirder // true
+
+val fiveWeirdNumbers = Set(0) *: evenWeirder // (Set(0),1,2,3.0,List(4))
+```
+
+When we use a tuple as argument of `*:`, it is prepended as a single element:
+```scala
+val notGood: ((Int, Int), Int, Int) = (1, 2) *: (3,4) // ((1, 2), 3, 4)
+```
+So how can we concatenate two tuples ?
+The `++` is there exactly for this purpose:
+```scala
+val better: (Int, Int, Int, Int) = (1, 2) ++ (3, 4) // (1, 2, 3, 4)
+```
+
+### Tuples can shrink
+
+Similarly to operators available on lists, we can retrieve a subset of
+a tuple. Here is a quick overview:
+
+ - `drop` allows to ignore the first *n* elements of the tuple, returning
+   an empty tuple when the number of elements is smaller than *n*:
+```scala
+(1, "2", 3.0, List(4)).drop(2) // (3.0, List(4))
+(1, "2", 3.0, List(4)).drop(10) // ()
+```
+ - `take` retrieves the first *n* elements of the tuple, returning the original
+   tuple when the number of elements is smaller than *n*
+```scala
+(1, "2", 3.0, List(4)).take(2) // (1, "2")
+(1, "2", 3.0, List(4)).take(10) // (1, "2", 3.0, List(4))
+```
+ - `splitAt` creates two tuples, the first of which contains the first *n* elements
+   of the original tuple and the second contains the remaining elements
+```scala
+(Set(0), 1, "2", 3.0, List(4)).splitAt(3) // ((Set(0), 1, "2", 3.0), (3.0, List(4)))
+```
+
+### Tuples can transform
+
+Again, similarly to conversion methods on collections, it is possible to
+transform a tuple into another collection.
+
+We have to pay attention to the type of the resulting collection.
+Let's start with the simple case: as its name might hint,
+`toArray` produces an array. The type of its elements will always be
+`Object`. This makes it easy to reason about this method although it
+forgets the type of the elements.
+It is also possible to use `.toIArray` which has exactly the same behavior
+but produces an `IArray` where the `I` stands for immutable.
+```scala
+scala> List(1, "2").toArray
+val res0: Array[Object] = Array(1, 2)
+```
+
+I believe however that the most interesting conversion is `toList`
+which produces yes a list, but not a `List[Any]`.
```

Comment:
That sounds weird. Consider dropping the "yes a list but not ...", and instead directly says it produces a `List[U]`.